### PR TITLE
Update the unification work required for the external package development

### PR DESCRIPTION
**Pull request by maliasadi**
This PR fixes the following issues.
1. The Lightning Monorepo doesn’t consider dependent external package development:
   - The core libraries are not interfaced and cannot be accessed
   - The simulator libraries are not interfaced and cannot be accessed.
2. Catalyst extends `StateVectorLQubit` by implementing a managed-like state-vector class with dynamic qubit management. To add this support, 
    -  The `num_qubits_` needs to be a "protected" class variable so that it can be updated dynamically in `StateVectorLQubitDynamic`. 
    -  The implementation of simulators’ algorithms, observables, measurements should be updated to support any concrete implementation of `StateVectorLQubit`.

[sc-43458]

### CODE REVIEW

File: pennylane_lightning/core/src/simulators/lightning_qubit/utils/LinearAlgebra.hpp
Comment (mlxd) [inline pennylane_lightning/core/src/simulators/lightning_qubit/utils/LinearAlgebra.hpp:88]: Did clang-format do this?

File: pennylane_lightning/core/src/simulators/lightning_qubit/utils/LinearAlgebra.hpp
Comment (mlxd) [inline pennylane_lightning/core/src/simulators/lightning_qubit/utils/LinearAlgebra.hpp:159]: As above?

**Review by mlxd — CHANGES_REQUESTED** [review]
Once the changes are added, I'll approve.

Also, the RTD builder is failing now. There is a PR needed to fix this open at #491 which will be necessary before merging.

File: Makefile
Comment (maliasadi) [reply Makefile:6]: @mlxd If no one is against adding Ninja to Make commands in the team, I can add Ninja build (we use Ninja in Catalyst). I added `NPROC` to make procs dev-configurable.

File: pennylane_lightning/core/src/simulators/lightning_qubit/utils/LinearAlgebra.hpp
Comment (maliasadi) [reply pennylane_lightning/core/src/simulators/lightning_qubit/utils/LinearAlgebra.hpp]: Different version of clang-format :)

File: pennylane_lightning/core/src/simulators/base/StateVectorBase.hpp
Comment (maliasadi) [reply pennylane_lightning/core/src/simulators/base/StateVectorBase.hpp]: @AmintorDusko suggestion to make `num_qubits_` protected instead of `private` works fine if you think this's a better approach than adding `setNumQubits` with the default throwing exception. I like the latter slightly  more than making a base class value protected.

Comment (maliasadi) [reply pennylane_lightning/core/src/simulators/base/StateVectorBase.hpp]: I leave the final decision to you guys (@AmintorDusko @mlxd) as I don't have any strong opinion on this. :100:

File: Makefile
Comment (mlxd) [reply Makefile:6]: I'd prepare to migrate away from the makefile too if you are depending on it, since it won't work on Windows.

File: pennylane_lightning/core/src/simulators/base/StateVectorBase.hpp
Comment (mlxd) [reply pennylane_lightning/core/src/simulators/base/StateVectorBase.hpp]: Since this is still hidden under the hood, I prefer not to add methods that aren't needed except for a single location's use. Fine with the move to protected here if it solves the issue.

We can always revisit a more wide-spread integration of the dynamic functionality later, but let's use this approach for now.

File: Makefile
Comment (maliasadi) [reply Makefile:6]: We don't use Make commands of Lighting repo in Catalyst. NPROC is simply a minor addition to make building/testing lightning faster and I can remove it, keep it as is or extend it to use ninja build-system.

File: pennylane_lightning/core/src/simulators/base/StateVectorBase.hpp
Comment (maliasadi) [reply pennylane_lightning/core/src/simulators/base/StateVectorBase.hpp]: Sounds good! Yes, the protected `num_qubits_` works too.

File: pennylane_lightning/core/src/simulators/lightning_qubit/algorithms/AdjointJacobianLQubit.hpp
Comment (AmintorDusko) [inline pennylane_lightning/core/src/simulators/lightning_qubit/algorithms/AdjointJacobianLQubit.hpp:259]: There is a reason for these ``if constexpr`` to be as they were. 
We cannot guarantee that a catch-all ``else`` will really cover all cases.
With these changes, if a developer provides a StateVector without internal storage (like Raw) we'll have a code-wise unexpected behaviour.

**Review by AmintorDusko — COMMENTED** [review]
Hi @maliasadi, I'm happy with having ``num_qubits`` changed to protected.
In terms of algorithms, measurements and observables, I would rather have an explicit reference/check with respect to the ``StateVectorLQubitCatalist`` than a catch-all that assumes that everything that is not ``StateVectorLQubitRaw`` will be ``StateVectorLQubitManaged``-like.
This brings us to another topic. In terms of all we discussed, and continuous support for ``StateVectorLQubitCatalist``, I think it would be a good call to have it added as a new ``StateVectorLQubit`` variant to Lightning. In this way, our CI/CD will make sure no change breaks it. For example, having ``num_qubits`` as private works very well for Lightning's needs so far, if we have a backend relying on it the team will always remember (or be remembered of) this constraint.

File: CMakeLists.txt
Comment (multiphaseCFD) [inline CMakeLists.txt:104]: Thanks @maliasadi! Could you please also change `pennylane_lightning/src` to `pennylane_lightning/core/src` here?

**Comment by mlxd** [comment]
Hey @maliasadi feel free to retag us for review once this is updated.

File: CMakeLists.txt
Comment (maliasadi) [reply CMakeLists.txt]: Thanks @multiphaseCFD! Fixed [a7fcf1d](https://github.com/PennyLaneAI/pennylane-lightning/pull/482/commits/a7fcf1dafe848417c8d535b2c838af709e732a98)

File: pennylane_lightning/core/src/simulators/lightning_qubit/algorithms/AdjointJacobianLQubit.hpp
Comment (maliasadi) [reply pennylane_lightning/core/src/simulators/lightning_qubit/algorithms/AdjointJacobianLQubit.hpp]: Thanks @AmintorDusko!

**Review by mlxd — APPROVED** [review]
Thanks @maliasadi 
Fine with me to see this go through if it meets your needs

**Review by AmintorDusko — APPROVED** [review]
Happy to have this in, if this will help PennyLane catalyst!

**Pull request merged**
